### ISSUE TITLE
CORS issue: Request header field x-msw-request-id is not allowed by Access-Control-Allow-Headers

### ISSUE THREAD

**Issue by romaolucas** [kettanaito/msw#1022]
## Environment

| Name    | Version |
| ------- | ------- |
| msw     | 0.36.2   |
| google chrome |  95.0.4638.69  |
| Ubuntu      | 20.0.4     |

## Request handlers

```js

const setup = () => {
  if (typeof window === 'undefined') {
    const { server } = require('./server');
    server.listen({ onUnhandledRequest: 'bypass' });
  } else {
    const { worker } = require('./browser');
    worker.start({ onUnhandledRequest: 'bypass' });
  }
};

export default setup;
```

Server setup
```js
import { rest } from 'msw';
import { setupServer } from 'msw/node';

const server = setupServer([
rest.get('*/house/:externalId/info', (req, res, ctx) =>
    res(ctx.json(mock.house)),
  ),
  rest.get('*/house-enums', (req, res, ctx) =>
    res(ctx.json(enumMock.houseEnums)),
  ),
  rest.get('*/bff/sales-flow/:salesFlowId/ccv-card', (req, res, ctx) =>
    res(ctx.json(ccvMock.validCCVData)),
  ),
  rest.patch('*/bff/sales-flow/:salesFlowId/ccv-card', (req, res, ctx) =>
    res(ctx.status(200)),
  ),
  rest.get('*/ccv-enums', (req, res, ctx) => res(ctx.json(ccvMock.ccvEnums))),
  rest.get('*/sales-flow/:salesFlowId/diligence', (req, res, ctx) =>
    res(ctx.json(diligenceResponseMock.diligenceData)),
  ),
]);

export { server, rest };

```

## Actual request

This request is a bypassed one made by an authentication library that we use internally.
The fetch to `https://sso.keycloak.com.br/auth/realms/forno/protocol/openid-connect/token` passes a code received in a previous request and a redirect url and expects the authentication token as a response.

## Current behavior

Receiving following error on console
```
Access to XMLHttpRequest at 'https://sso.keycloak.com.br/auth/realms/forno/protocol/openid-connect/token' from origin 'http://vendas.dev.quintoandar.com.br' has been blocked by CORS policy: Request header field x-msw-request-id is not allowed by Access-Control-Allow-Headers in preflight response.
```

## Expected behavior

Request is successfully bypassed without CORS issue.
Is there a way to exclude this header for bypassed requests?

**Comment by kettanaito** [comment]
Hey, @romaolucas. Thank you for reporting this.

## Root cause for the exception

The issue is caused by the CORS policy of your server not accepting custom `x-` request headers. Due to that, the request is rejected because it violates the configured CORS policy (`Access-Control-Allow-Headers` in particular).

## Root cause for the request header

Historically, the library appended the `x-msw-request-id` header to associate responses with the corresponding requests in Node.js (see #713). Now, it seems that we can fully rely on the interceptors library we're using to provide us with the necessary properties for response/request mapping (`request.id`). 

I'm removing the `x-msw-request-id` header from the request transformations in #1024, which should fix your issue.

**Comment by kettanaito** [comment]
Please update to [0.36.3](https://github.com/mswjs/msw/releases/tag/v0.36.3) to have the fix.

**Comment by romaolucas** [comment]
Thanks Artem, I'll do that!

Em qui., 9 de dez. de 2021 às 11:53, Artem Zakharchenko <
***@***.***> escreveu:

**Comment by awacode21** [comment]
@kettanaito i am still having that problem with version 0.36.7

**Comment by kettanaito** [comment]
Hey, @awacode21. Could you please update to the latest `msw` version? You can do that by `npm i msw@latest -D`. Right now it's 0.39.2. We've migrated from using `x-msw-request-id` request header to track requests.

**Comment by awacode21** [comment]
hi @kettanaito updated to latest version. Working on my application. but not working when using MSW for storybook. so maybe it is related to the msw-storybook-addon that there is a problem?

**Comment by kettanaito** [comment]
Oh, so that's the reason. We've had to bump the `peerDependency` range in `msw-storybook-addon` just recently (https://github.com/mswjs/msw-storybook-addon/issues/74). Please update it as well to have the fix propagate to your project:

```sh
npm i msw-storybook-addon@latest
```

Also, if you run `npm ls msw` you will see what packages in your project depend on `msw`. It'd have helped you to spot that the Storybook add-on installs a much older version of `msw` (0.35), and both 0.39 and 0.35 end up being used in your app.

**Comment by awacode21** [comment]
@kettanaito i already updated the msw-storybook-addon a few minutes ago, but the error still remains.

i tried the command you postet, and got the following result:

```
awacode@awacode-MBP13 EKP-Frontend % npm ls msw
d-ekp@0.0.0 /Users/awacode/code/customers/enbw/EKP-Frontend
├─┬ msw-storybook-addon@1.6.3
│ └── msw@0.39.2 deduped
└── msw@0.39.2
```

Seem to use the right version.

**Comment by awacode21** [comment]
<img width="607" alt="Bildschirmfoto 2022-04-09 um 21 18 50" src="https://user-images.githubusercontent.com/30939926/162588666-c3fc6803-f8f5-4d31-b311-a6c07a69f11b.png">

**Comment by kettanaito** [comment]
I see. This doesn't look like the error related to the `x-msw-request-id` issue. We'd see that header being rejected by the server. What we see in the error above is that endpoint X doesn't allow `localhost` like an origin. Sorry, I don't have enough context to suggest anything meaningful right now. I'd ask you to create a reproduction repository for this and submit a new issue. 

I can only propose to double-check that MSW is enabled in Storybook (you see the console log message). Double-check with the add-on docs to be sure.

**Comment by awacode21** [comment]
@kettanaito thank you for all your ideas, your last post gave me the important hint. the request got bypassed by msw because it found no match.

i did the msw handler configuration globally in preview.js and this seem not to work. when i am doing the exact same configuration directly as parameters of a specific story, it works.

so it looks like global msw parameter configuration in preview.js somehow is broken, whereas configuring it for a specific story is fine. will open a new issue.

**Comment by kettanaito** [comment]
Oh, wow, that's unexpected. I wonder what in our tests makes us miss this issue. Please, may I ask you to submit it to the [add-on repository](https://github.com/mswjs/msw-storybook-addon)? Thanks.

**Comment by awacode21** [comment]
Figured out it was related to a misconfiguration of the whole parameters object. Related to storybook itself. So luckily it is not related to MSW itself.

**Comment by mrosata** [comment]
> Figured out it was related to a misconfiguration of the whole parameters object. Related to storybook itself. So luckily it is not related to MSW itself.

@awacode21 I know this was almost a year ago, but if you have time could you post your fix to help others in the same situation? (namely me 😄 , what was the fix to global parameter mocks)

---
**EDIT:** Figured it out. For future reference, anyone who is having similar issues where global mocks do not work correctly with individual story mocks, it seems `pararameters.msw` will take either an array or an object where `handlers.<key>` is an array of mock handlers. For complex mocking scenarios the nested object works better.
For example:
```ts
// preview.jsx
export parameters = {
  msw: {
    handlers: {
      metrics: [
        rest.post("*/report/metric", (_, res, ctx) =>
          res(ctx.status(200), ctx.json({}))
        ),
      ],
    },
  },
};
```

```ts
// ../some-component/storybook-helpers/requests.ts
export parameters = {
  msw: {
    handlers: {
      mocks: [
        rest.post("*/api/peace/love", (_, res, ctx) =>
          res(ctx.status(200), ctx.json({}))
        ),
      ],
    },
  },
};
```
Storybook will properly merge the global `parameters.msw.handlers` and some-components `parameters.msw.handlers`. Because they have different keys (`metrics` and `mocks`), the merge doesn't clobber either of them and they will live and play in peace and harmony ☮️
